commons/src/fun_run.rs

```diff
@@ -0,0 +1,280 @@
+/// - Return error messages with the command name.
+/// - Turn non-zero status results into an error
+/// - Embed stdout and stderr into errors (when not streamed)
+/// - Store stdout and stderr without displaying them (when streamed)
+///
+/// Even better:
+///
+/// - Composable by design. Use what you want. Ignore what you don't.
+/// - Plays well with standard library types by default.
+///
+/// And of course:
+///
+/// - Fun(ctional)
+///
+/// The main interface is the `name_and_then` method provided by the `CommandNameAndThen` extension.
+/// Use this along with other fun methods to compose the command run of your dreams.
+///
+/// Example:
+///
+/// ```no_run
+/// use commons::fun_run::{ self, CommandNameAndThen};
+/// use libherokubuildpack::command::CommandExt;
+/// use std::process::Command;
+/// use libcnb::Env;
+///
+/// let env = Env::new();
+///
+/// Command::new("bundle")
+///     .args(["install"])
+///     .envs(&env)
+///     .name_and_then(fun_run::display, |name, cmd| {
+///         eprintln!("\nRunning command:\n$ {name}");
+///
+///         cmd.output_and_write_streams(std::io::stdout(), std::io::stderr())
+///             .map_err(|error| fun_run::on_system_error(name.clone(), error))
+///             .and_then(|output| fun_run::nonzero_streamed(name.clone(), output))
+///     }).unwrap();
+/// ```
+
+/// Format a command in a single location use it for display or errors
+pub trait CommandNameAndThen<N, G, O, E>
+where
+    N: Fn(&mut Command) -> String,
+    G: FnOnce(String, &mut Command) -> Result<O, E>,
+{
+    /// Takes two functions the first produces a name
+    /// and is then passed to the section function along with
+    /// an `&mut Command`
+    ///
+    /// The second function is intended to return
+    /// `Result<Output, std::io::Error>` or `Result<Output, CmdError>`, but can
+    /// return any result.
+    ///
+    /// # Errors
+    ///
+    /// Returns the result of the second function directly.
+    ///
+    fn name_and_then(&mut self, get_name: N, f: G) -> Result<O, E>;
+}
+
+impl<N, G, O, E> CommandNameAndThen<N, G, O, E> for Command
+where
+    N: Fn(&mut Command) -> String,
+    G: FnOnce(String, &mut Command) -> Result<O, E>,
+{
+    fn name_and_then(&mut self, get_name: N, f: G) -> Result<O, E> {
+        let name = get_name(self);
+        f(name, self)
+    }
+}
+
+lazy_static! {
+    // https://github.com/jimmycuadra/rust-shellwords/blob/d23b853a850ceec358a4137d5e520b067ddb7abc/src/lib.rs#L23
+    static ref QUOTE_ARG_RE: regex::Regex = regex::Regex::new(r"([^A-Za-z0-9_\-.,:/@\n])").expect("Internal error:");
+}
+
+/// Converts a command and its arguments into a user readable string
+///
+/// Example
+///
+/// ```rust
+/// use std::process::Command;
+/// use commons::fun_run;
+///
+/// let name = fun_run::display(Command::new("bundle").arg("install"));
+/// assert_eq!(String::from("bundle install"), name);
+/// ```
+#[must_use]
+pub fn display(command: &mut Command) -> String {
```

Comment:
Why not implement `Display` for `Command`? That way, you could `Command::new("bundle").arg("install").to_string()`.

Reply:
We can't implement `Display` directly due to https://github.com/Ixrec/rust-orphan-rules. There's a way to do it that's hacky that involves creating our own trait that then implements that trait, but it's well...hacky.

I'm not opposed to that existing, I just also think people want to have the option of using a method if they want. 

The other concern is different people need different formatting. I want/need to have env vars in my displays.